include/sycl_ext_complex.hpp

```diff
@@ -1679,6 +1707,497 @@ _SYCL_EXT_CPLX_INLINE_VISIBILITY
+template <typename BinaryOperation>
+inline constexpr bool is_multiplies_v = is_multiplies<BinaryOperation>::value;
+
+/// Wrapper trait to check if the binary operation is supported
+template <typename BinaryOperation>
+struct is_binary_op_supported
+    : std::integral_constant<bool,
+                             (detail::is_plus<BinaryOperation>::value ||
+                              detail::is_multiplies<BinaryOperation>::value)> {
+};
+template <class BinaryOperation>
+inline constexpr bool is_binary_op_supported_v =
+    is_binary_op_supported<BinaryOperation>::value;
+
+/// Helper functions to get the init for sycl::plus binary operation when the
+/// type is a gencomplex
+template <typename T, class BinaryOperation>
+std::enable_if_t<(sycl::ext::cplx::is_gencomplex_v<T> &&
+                  detail::is_plus_v<BinaryOperation>),
+                 T>
+get_init() {
+  return T{0, 0};
+}
+/// Helper functions to get the init for sycl::multiply binary operation when
+/// the type is a gencomplex
+template <typename T, class BinaryOperation>
+std::enable_if_t<(sycl::ext::cplx::is_gencomplex_v<T> &&
+                  detail::is_multiplies<BinaryOperation>::value),
+                 T>
+get_init() {
+  return T{1, 0};
+}
+/// Helper functions to get the init for sycl::plus binary operation when the
+/// type is a mgencomplex
+template <typename T, class BinaryOperation>
+std::enable_if_t<
+    (is_mgencomplex_v<T> && detail::is_plus<BinaryOperation>::value), T>
+get_init() {
+  using Complex = typename T::value_type;
+
+  T result;
+  std::fill(result.begin(), result.end(), Complex{0, 0});
+  return result;
+}
+/// Helper functions to get the init for sycl::multiply binary operation when
+/// the type is a mgencomplex
+template <typename T, class BinaryOperation>
+std::enable_if_t<
+    (is_mgencomplex_v<T> && detail::is_multiplies<BinaryOperation>::value), T>
+get_init() {
+  using Complex = typename T::value_type;
+
+  T result;
+  std::fill(result.begin(), result.end(), Complex{1, 0});
+  return result;
+}
+
+} // namespace cplex::detail
+
+/* REDUCE_OVER_GROUP'S OVERLOADS */
+
+/// Complex specialization
+template <typename Group, typename V, typename T, class BinaryOperation,
+          typename = std::enable_if_t<
+              sycl::is_group_v<std::decay_t<Group>> && is_genfloat_v<V> &&
+              is_genfloat_v<T> &&
+              cplex::detail::is_binary_op_supported_v<BinaryOperation>>>
+complex<T> reduce_over_group(Group g, complex<V> x, complex<T> init,
+                             BinaryOperation binary_op) {
+#ifdef __SYCL_DEVICE_ONLY__
+  complex<T> result;
+
+  result.real(sycl::reduce_over_group(g, x.real(), init.real(), binary_op));
+  result.imag(sycl::reduce_over_group(g, x.imag(), init.imag(), binary_op));
+
+  return result;
+#else
+  throw sycl::exception(sycl::make_error_code(sycl::errc::runtime),
+                        "Group algorithms are not supported on host.");
+#endif
```

Comment:
This is horrendously unportable.
* The spec defines `__SYCL_DEVICE_ONLY__` only for SMCP compilers. For single-pass compilers, like the hipSYCL generic SSCP compiler,  this is not defined in accordance with the SYCL spec
* Just throwing whenever `__SYCL_DEVICE_ONLY__` is not defined will completely break on any library-only host backend, where the kernel is compiled as part of the host pass for CPU.

You should seriously start looking into adding CI and validation for other compilers. That would make such issues blatantly obvious.

Reply:
I'd like to add openSYCL CI - do you have a public docker image that we could use with latest version pre-installed?

Re avoiding the `__SYCL_DEVICE_ONLY__` else raise pattern, do you have any suggestions on an alternative? This has come up in gtensor as well, where we have a function that could be called from host or device code, and want to have some error reporting in both cases. Is there a standard SYCL safe way to always report an error for both device and host calls?

Reply:
We don't have a docker image at the moment (it's on the lower priority to do list to speed up our own CI), but what you need is usually fairly simple:
1. Install some recent LLVM release from `apt.llvm.org` packages, boost packages, cmake, git
2. Clone and install.

See our own CI (which is a bit more complicated because it tests all backends in different versions etc): https://github.com/OpenSYCL/OpenSYCL/blob/develop/.github/workflows/linux.yml



> Re avoiding the __SYCL_DEVICE_ONLY__ else raise pattern, do you have any suggestions on an alternative? This has come up in gtensor as well, where we have a function that could be called from host or device code, and want to have some error reporting in both cases. Is there a standard SYCL safe way to always report an error for both device and host calls?

I might be able to answer more specifically with more information about your use case, but in general this seems like a difficult/impossible thing to do:
* There is currently no portable mechanism in SYCL to specialize host/device code.
* As soon as a function is in the kernel call graph, it counts as device code and device code restrictions in principle apply (no exceptions are allowed or else UB). This also affects the host side, as SYCL implementations might use that to target CPU, and still rely on the absence of exceptions for optimizations.
* Error handling from within kernels is always very difficult, and there is no mechanism for this in the standard, unless you are willing to pass `sycl::stream` everywhere throughout your code (you might not want that, as its bare existence might add overhead depending on the implementation). Some implementations may support printf/assert but it's not guaranteed.

The issue for pre-standardized APIs like the SYCL group algorithms is more difficult because we cannot add to the API. Outside of this, one suggestion could be to modify the API and solve this on the C++ level and not on the SYCL level using different template instantiations or overloads to distinguish host and device code paths.
In a single-pass compilation scenario, host and device are by nature extremely closely related and potentially generated from the exact same IR (even though we have mechanisms to distinguish them, but they are very different from the macro), thus they may not even be a separate thing a priori.

Can you describe your use case a bit more - how your error handling is intended to work and why you feel the need for raising errors inside the kernel call graph?